doc/proposals/openshift-4.3/operator-testing-tool.md

```diff
@@ -0,0 +1,240 @@
+
+The scorecard would by default run all tests regardless if a single test fails.  Using a CLI flag such as below would cause the test execution to stop on a failure:
+ * operator-sdk scorecard -l ‘testgroup=required’ --fail-fast
+
+Tasks for this story include:
+ * change scorecard by removing earnedPoints and maximumPoints from each test
+ * change scorecard output by removing totalScorePercent, partialPass, earnedPoints, maximumPoints from scorecard output
+ * change scorecard existing tests to remove partialPass, instead only returning a pass or fail
+ * change scorecard CLI to return an exit code based on the pass/fail results from the set of executed tests 
+ * (stretch goal) add support for --fast-fail CLI flag
+
+
+Rough point estimate is 8 points.
+
+
+#### Story 2 - Change Scorecard Test Selection
+Today, the scorecard lets you select which tests are run by the plugins defined in the scorecard configuration (basic, olm). This story would change how users determine which tests are executed by scorecard.
+
+This story would introduce labels for each test and then allow the scorecard CLI to filter which tests it runs based on label selectors.
+
+Tests can fall into 3 groups: required, recommended, or optional. Tests can also be categorized as static or runtime. Labels for these groups and categories would allow a test to be more precisely specified by an end user.
+
+Possible examples of specifying what tests to run are as follows: 
+ * operator-sdk scorecard -l ‘testgroup in (required,recommended,optional)’ 
+ * operator-sdk scorecard -l ‘testtype in (runtime,static)’
+ * operator-sdk scorecard -l updatecrtest=true
+ * operator-sdk scorecard -l someplugintest=true
+
+Tasks for this story include:
+ * Update the test interfaces to support arbitrary test-defined labels
+ * Add the labels to the tests
+ * Update the CLI and scorecard library to support label selectors to filter scorecard runs
+ * A stretch goal would include the ability of the CLI to perform a dry-run of tests based on label selectors to show the user which tests would be executed for a given label selector value.
+
+Rough point estimate for this story is 8 points.
+#### Story 3 - Common Validation
+The scorecard would use a common validation codebase to verify bundle contents. This story supports a single “source of truth” and avoids duplication or variance of validation logic. The scorecard user would be able to specify a bundle on the command line, that bundle would then be validated.  For example:
+ * operator-sdk scorecard --bundle ./my-operator
+
+
+Tasks for this story include:
+ * determine the correct validation library to use
+ * add support for a bundle directory flag to the CLI
+ * change the scorecard code to use that validation library when validating a CR, CSV, or given a bundle directory
```

Comment:
IMHO: It should be:

- CR/CRD/RBCA -> By default
- OLM files (CSV) are optional - IMHO should be checked if they exist in the project and/or a flag for it be used.

Note: The bundle in POV is not an intuitive nomenclature.

Reply:
I think one issue with the scorecard today is that it has too much configuration complexity. The following flags can all be used to define how scorecard runs. Some are optional, some are required, and some are conditionally required. There are also interdependencies between them that can cause confusion.
```
      --basic-tests                  Enable basic operator checks (default true)
      --cr-manifest strings          Path to manifest for Custom Resource (required) (specify flag multiple times for multiple CRs)
      --crds-dir string              Directory containing CRDs (all CRD manifest filenames must have the suffix 'crd.yaml') (default "deploy/crds")
      --csv-path string              Path to CSV being tested
      --global-manifest string       Path to manifest for Global resources (e.g. CRD manifests)
      --namespace string             Namespace of custom resource created in cluster
      --namespaced-manifest string   Path to manifest for namespaced resources (e.g. RBAC and Operator manifest)
      --olm-deployed                 The OLM has deployed the operator. Use only the CSV for test data
      --olm-tests                    Enable OLM integration checks (default true)
```

If we go all-in on OLM integration (which I believe is a goal), then I think we should require a bundle and leverage other OLM integrations (e.g. installing OLM and running an operator with OLM) so that we can reduce configuration complexity and make running and troubleshooting scorecard much more straightforward.

@dmesser @shawn-hurley WDYT?